File: BinaryChop/python/binarychop.py
import sys
import time


def binary_chop(number: int, arraysearch: list) -> int:
    left = 0
    right = len(arraysearch) - 1
    middle = (right + left) // 2
    found = -2
    while found == -2:
        if arraysearch[middle] < number:
            left = middle + 1
            middle = (right + left) // 2
        elif arraysearch[middle] > number:
            right = middle - 1
            middle = (right + left) // 2
        elif arraysearch[middle] == number:
            return middle
        elif left == right:
            found = -1
    return found


def binary_chop_recursive(number: int, arraysearch: list, left: int, right: int) -> int:
    if left == right:
        return -1
    middle = (right + left) // 2
    if arraysearch[middle] == number:
        return middle
    if arraysearch[middle] < number:
        left = left + 1
    if arraysearch[middle] > number:
        right = right - 1
    return binary_chop_recursive(number, arraysearch, left, right)


array = [i for i in range(10000)]
start_time_iterative = time.time()
print('El resultado es ', format(binary_chop(0, array)))
print(f'Execution time iterative {time.time() - start_time_iterative}')
sys.setrecursionlimit(100000)
start_time_recursive = time.time()
print('El resultado recursivo es ', format(binary_chop_recursive(6, array, 0, len(array))))
print(f'Execution time recursive {time.time() - start_time_iterative}')

File: BinaryChop/kotlin/src/BinaryChop.kt
class BinaryChop {

    fun binaryChop(number: Int, array: IntArray): Int
    {
        var left = 0
        var right = array.size
        var middle = (right+left)/2
        var found = -2
        while(found == -2) {
            if (array[middle] < number){
                left = middle + 1
                middle = (right + left) /2
            }
            if(array[middle] > number){
                right = middle - 1
                middle = (right + left) / 2
            }
            if(array[middle] == number){
                found = middle
            }
            if(left > right) {
                return -1
            }
        }
        return found
    }

    fun binaryChopRecursive(number: Int, array: IntArray, left: Int, right: Int): Int
    {
        if(left > right){
            return -1
        }
        val middle = (right+left) / 2
        if (array[middle] == number ){
            return middle;
        }
        var newLeft = left;
        var newRight = right;
        if(array[middle] < number ){
            newLeft = middle + 1
        }
        if(array[middle] > number) {
            newRight = middle - 1
        }
        return binaryChopRecursive(number, array, newLeft, newRight)
    }
}